### PR TITLE
Merge helix-vm-freeze feature branch to master branch

### LINKED ISSUE #2018: Helix needs a cloud event management framework
**Is your feature request related to a problem? Please describe.**
Helix has added support features for cloud such as auto registration. But it is missing a mechanism to enable users to easily handle cloud events, for example cloud maintenance event. 

**Describe the solution you'd like**
Helix can develop a framework which based on listeners to enable users to plug in logics to handle cloud events.

**Additional context**

### PR DESCRIPTION
**Pull request by mgao0**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes  #2018

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This commit adds new concepts:
1. **CloudEventListener**: It represents an event listener, user will need to choose callback from what is available and register their own callback logic to event listener for different types of cloud event.
2. **CloudEventHandler**: It is fundamentally a list of cloud listeners, and is meant to provide convenience for users for registering listeners. Within Helix usage, users can register multiple event listeners to the event handler. So when registering listeners to the cloud platform, users just need to register this CloudEventHandler object, which means the listeners that are previously registered to this event handler are registered  to the cloud platform all at once.

This commits also embeds the cloud event handling logic into HelixManager. More specifically, ZKHelixManager can be populated with a cloud event listener if user defines it in **CloudEventCallbackProperty** which is nested inside HelixCloudProperty and HelixManagerProperty. CloudEventCallbackProperty not only allows users to choose from Helix-provided callback methods, as well as override them with user specific logics, but also users can supply their callback logic and specify the order that the callbacks will be triggered (before  or after Helix-supported callback).

### Tests

- [x] The following tests are written for this issue:

TestCloudEventCallbackProperty
TestDefaultCloudEventCallbackImpl

- The following is the result of the "mvn test" command on the appropriate module:

**CI all tests passed.**

Results from local run:
[INFO] Tests run: 1307, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6,442.479 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1307, Failures: 0, Errors: 0, Skipped: 0

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Review by qqu0127 — APPROVED** [review]
LGTM, assuming all patches have been reviewed earlier. Please also have a committer review too. thx

**Comment by mgao0** [comment]
This PR is ready to be merged. Approved by @qqu0127 , thanks.

Commit message:
Merge helix-vm-freeze feature branch to master branch, change includes #1976, #1978, #2017, and #1995.

**Pull request merged**
